File: src/main/java/com/devmpv/repositories/AttachmentRepository.java
package com.devmpv.repositories;

import java.io.Serializable;
import java.util.Set;

import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.CrudRepository;

import com.devmpv.model.Attachment;

/**
 * Attachment repository.
 * 
 * @author devmpv
 *
 */
public interface AttachmentRepository extends CrudRepository<Attachment, Serializable> {

    @Query("select count(m) from Attachment a join a.messages m where a.id = ?1")
    Long countMessages(long id);

    Attachment findByMd5(String md5);

    @Query("select a from Attachment a where a.messages.size=0")
    Set<Attachment> findEmpty();
}

File: src/main/kotlin/com/devmpv/repositories/AttachmentRepository.kt
package com.devmpv.repositories

import com.devmpv.model.Attachment
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.CrudRepository
import java.io.Serializable
import java.util.*

/**
 * Attachment repository.
 *
 * @author devmpv
 */
interface AttachmentRepository : CrudRepository<Attachment, Serializable> {

    @Query("select count(m) from Attachment a join a.messages m where a.id = ?1")
    fun countMessages(id: Long): Long?

    fun findByMd5(md5: String): Optional<Attachment>

    @Query("select a from Attachment a where a.messages IS EMPTY")
    fun findEmpty(): Set<Attachment>
}
